### PR TITLE
Add reentrant option for acquireLock()

### PR DESCRIPTION
**Pull request by lcabancla**
*Issue #, if available:* https://github.com/awslabs/amazon-dynamodb-lock-client/issues/56

*Description of changes:*

Add a reentrant option for acquireLock(). With this set to true, the lock client will check first if it already owns the lock. If it already owns the lock and the lock is not expired, it will return the lock immediately. If this is set to false and the client already owns the lock, the call to acquireLock will block.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/amazonaws/services/dynamodbv2/AcquireLockOptions.java
Comment (KasulaVamsi) [inline src/main/java/com/amazonaws/services/dynamodbv2/AcquireLockOptions.java:336]:
```suggestion
                    this.updateExistingLockRecord, this.shouldSkipBlockingWait, this.acquireReleasedLocksConsistently, this.reentrant);
```

- consistency

Comment (KasulaVamsi) [reply src/main/java/com/amazonaws/services/dynamodbv2/AcquireLockOptions.java:336]: Also update the toString() below

Comment (lcabancla) [reply src/main/java/com/amazonaws/services/dynamodbv2/AcquireLockOptions.java]: Addressed review comments in latest commit.

File: src/main/java/com/amazonaws/services/dynamodbv2/AmazonDynamoDBLockClient.java
Comment (lcabancla) [reply src/main/java/com/amazonaws/services/dynamodbv2/AmazonDynamoDBLockClient.java:577]: `hasLock()` is essentially a simplified version of `[getLock()](https://github.com/awslabs/amazon-dynamodb-lock-client/blob/master/src/main/java/com/amazonaws/services/dynamodbv2/AmazonDynamoDBLockClient.java#L929)`  where, instead of returning an expired/unexpired lock it returns a boolean. 

Since getLock() does not check the owner I assume it is fine no to do so.

File: src/main/java/com/amazonaws/services/dynamodbv2/AmazonDynamoDBLockClient.java
Comment (lcabancla) [reply src/main/java/com/amazonaws/services/dynamodbv2/AmazonDynamoDBLockClient.java:575]: I would argue it should be public since it provides a clearer API for determining lock ownership compared to `getLock()`.

File: src/main/java/com/amazonaws/services/dynamodbv2/AcquireLockOptions.java
Comment (lcabancla) [reply src/main/java/com/amazonaws/services/dynamodbv2/AcquireLockOptions.java:344]: Will update.

File: src/main/java/com/amazonaws/services/dynamodbv2/AmazonDynamoDBLockClient.java
Comment (lcabancla) [reply src/main/java/com/amazonaws/services/dynamodbv2/AmazonDynamoDBLockClient.java:411]: Added new test in the latest commit.

File: src/main/java/com/amazonaws/services/dynamodbv2/AmazonDynamoDBLockClient.java
Comment (kapilsingh5050) [reply src/main/java/com/amazonaws/services/dynamodbv2/AmazonDynamoDBLockClient.java:575]: I don't have a strong opinion here but I prefer keeping public APIs/contracts to minimal by exposing only what's needed to outside classes

File: src/test/java/com/amazonaws/services/dynamodbv2/AmazonDynamoDBLockClientTest.java
Comment (kapilsingh5050) [inline src/test/java/com/amazonaws/services/dynamodbv2/AmazonDynamoDBLockClientTest.java:239]: Please make sure we have a test where the second call to acquire lock fails. If it's not there already, please add it

**Review by kapilsingh5050 — APPROVED** [review]
Approved with a few comments. Please address the comment regarding test in AmazonDynamoDBLockClientTest.java file before merging

**Review by KasulaVamsi — APPROVED** [review]

**Comment by lcabancla** [comment]
@kapilsingh5050 Thanks for the approval. Added a test. 
I don't have write access to merge. Also, I am not familiar with the process of releasing a version. Do you know how to do that?

Screen Shot 2021-06-28 at 11 30 38 AM

**Comment by kapilsingh5050** [comment]
I also don't have access to merge these PRs. I'm also not familiar with the release process. @KasulaVamsi Can you check if you have permissions to merge?

**Pull request merged**
